fem/pbilinearform.hpp

```diff
@@ -150,6 +150,27 @@ class ParBilinearForm : public BilinearForm
        vectors on the true dofs. */
    void TrueAddMult(const Vector &x, Vector &y, const double a = 1.0) const;
 
+   /// Compute \f$ y^T M x \f$
+   /** @warning The calculation is performed on local dofs, assuming that
+       the local vectors are consistent with the prolongations of the true
+       vectors (see ParGridFunction::Distribute()). If this is not the case,
+       use TrueInnerProduct(const ParGridFunction &, const ParGridFunction &)
+       instead. */
+   double InnerProduct(const ParGridFunction &x,
+                       const ParGridFunction &y) const;
```

Comment:
Here is an issue I came across while testing this, though this is an existing issue not introduced by this PR. 
`ParBilinearForm::InnerProduct` calls `BilinearForm::InnerProduct` and this in turn calls `SparseMatrix::InnerProduct` for the internal `SparseMatrix * mat`. In case 
 `ParBilinearForm::FormSystemMatrix` was called before, then it segfaults because  `mat` had been deleted. For the same reason,  all `[Par]BilinearForm::[True]InnerProduct` calls cannot work with `-pa`.  I think this shouldn't affect this PR, but it has to be fixed either by adding a check, or we can simply implement it in a separate PR. I'll wait to see what others think.

Reply:
I added an assert and a note that the local matrix is assumed to be assembled, because the situation is different compared to `BilinearForm`, where the matrix is not deleted normally and it is a (nearly) clear mistake of the user to call operations on non-assembled form (of course an assert would be better than a segfault, but that is another matter...). In contrast, `ParBilinearForm` "replaces" the local matrix by the parallel during normal operation, when that `FormSystemMatrix()` is called, so it is reasonable to put an assert here.
Regarding operator-based forms (partial/element assembly/matrix free), that is another thing and there are many methods like that in `BilinearForm`, operating only on `SparseMatrix`. This should be solved on the side of `BilinearForm` and gets beyond the scope of this PR.